### ISSUE TITLE
S3AsyncClient hangs on many concurrent putObject operations. Logs "NoSuchElementException: RemoveAfterReadTimeoutHandler"

### ISSUE THREAD

**Issue by henriknume** [aws/aws-sdk-java-v2#1167]
## Current Behavior

When many putObject requests are performed concurrently the S3AsyncClient sporadically hangs/timesout when waiting for a list of CompletableFutures to finish.  The number of requests required for the bug to appear varies greatly but N>50 in a lambda and N > 500 on a local machine usually gives the error.  The files being uploaded are all between 1-2kb in size.

The following error message is shown in logs: 
```
[aws-java-sdk-NettyEventLoop-1-7] WARN software.amazon.awssdk.http.nio.netty.internal.UnusedChannelExceptionHandler - A non-I/O exception occurred on a channel (dc9ab55d) that was not in use. The channel has been closed to prevent any ongoing issues.
java.util.NoSuchElementException: RemoveAfterReadTimeoutHandler
	at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:1084)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:427)
	at software.amazon.awssdk.http.nio.netty.internal.OneTimeReadTimeoutHandler.channelRead0(OneTimeReadTimeoutHandler.java:46)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at java.lang.Thread.run(Thread.java:748)
```

## Steps to Reproduce (for bugs)

In the following example snippet I'm using adobe s3mock but the same problem is observed  when running the code from a lambda to a real s3 bucket with similar code. 

 ```
 @Test
  public void test_client_with_local_s3_server() throws Exception {
    S3MockApplication s3 = S3MockApplication.start();

    S3AsyncClient s3Client = S3AsyncClient.builder()
        .credentialsProvider(
            StaticCredentialsProvider.create(AwsBasicCredentials.create("X", "X")))
        .endpointOverride(new URI("http://127.0.0.1:9090"))
        .region(Region.US_EAST_1)
        .build();

    // create bucket
    final String bucketName = "the-bucket-name";
    s3Client.createBucket(b -> b.bucket(bucketName)).get(3, SECONDS);
    List<String> objectKeyList = s3Client.listObjectsV2(b -> b.bucket(bucketName)).get().contents().stream().map(S3Object::key).collect(Collectors.toList());
    Assert.assertEquals(0, objectKeyList.size());

    // put list with N items
    int N = 500;
    List<String> fileNames = generateFileNames(N);

    CompletableFuture.allOf(
        fileNames.stream()
            .map(name -> s3Client.putObject(b -> b.bucket(bucketName).key(name), AsyncRequestBody.empty()))
            .toArray(CompletableFuture[]::new)
    ).get(15, SECONDS);

    // check list
    objectKeyList = s3Client.listObjectsV2(b -> b.bucket(bucketName)).get(3, SECONDS).contents().stream().map(S3Object::key).collect(Collectors.toList());
    System.out.println("Saved: " + objectKeyList.size());
    Assert.assertEquals(N, objectKeyList.size());
    s3.stop();
  }
```

## Your Environment

* AWS Java SDK version used: 2.5.16
* JDK version used: lambda runtime: java8  (test: 1.8.0_191)
* Operating System and version: lamba (test: Win10)

**Comment by dagnir** [comment]
@henriknume Thanks for the detailed bug report! I'll have go at trying to reproduce this on my machine today.

**Comment by zoewangg** [comment]
Hi @henriknume, I tried both locally (against s3 and mock server) and on Lambda and could not reproduce the`NoSuchElementException`.

What are your lambda memory and timeout settings? 
I noticed you didn't configure the [maxConcurrency](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L277) on the client and the default number is 50. Can you update it to match the number of the concurrent requests and try again? Though we still want to figure out why `NoSuchElementException: RemoveAfterReadTimeoutHandler` was thrown.

```
S3AsyncClient.builder()                                 
.httpClientBuilder(NettyNioAsyncHttpClient.builder().maxConcurrency(500))
.build();
```

**Comment by henriknume** [comment]
Lambda memory 3000mb (~200mb used)
maxConcurrency 100
Lambda timeout: 1min
get() timeout: 50s

I have played around with different client settings without any success. The exception is logged and not thrown by the client so it will still work most of the time, so I don't actually know if its related to the cause of  the problem. 

The problem is that some of the time one or a few of the concurrent requests will not complete ( lambda success-rate approx 99.8%). When I inspect the s3 requests in a failing lambda I can see that for example 49 of the 50 putObject requests are completed really quick maybe 30ms and then 1 requests gets stuck until the timeout of 50s is reached.  

When I use the non async client in the lambda it works 100% of the time but it obviously isn't as quick.

**Comment by dagnir** [comment]
Should be fixed by https://github.com/aws/aws-sdk-java-v2/pull/1235

**Comment by dagnir** [comment]
A bugfix for this has been released as part of `2.5.36`: https://github.com/aws/aws-sdk-java-v2/blob/2.5.36/CHANGELOG.md#netty-nio-http-client Please try it out and let us know if it resolves the issue.

**Comment by henriknume** [comment]
It looks like it's working now. Thanks!
